### PR TITLE
refactor cert-gen logic

### PR DESCRIPTION
**Pull request by dungdm93**
This is third part of refactoring helm chart #16792 aim to refactor cert-gen logic.

Currently, the cilium chart support 3 different types of cert-gen: provided, helm-gen, cronjob-gen. But the logic between those 3 types is not the same, for example: helm-gen method always generate both CA and certs (ignore user provided), but cronjob-gen reuse and disable generate cert when user provide CA/cert (this mean no cert auto-rotate).
https://github.com/cilium/cilium/blob/703b38fd47da1f91a342a8546f53542d624d8b22/install/kubernetes/cilium/templates/_hubble-generate-certs-job-spec.tpl#L47-L53  
So I simplify and unify this logic as following:
* Split each tls method to its own folder:
Screenshot from 2021-07-02 10-45-30
* When certificate is provided (`not auto.enabled`), both `cert` and `key` are required
* When certificate `auto.enabled` (both helm and cronjob methods), if user provide CA, use it; else generate new one. Other client/server/remote TLS are generated.

Signed-off-by: Đặng Minh Dũng <dungdm93@live.com>

### CODE REVIEW

File: install/kubernetes/cilium/templates/_helpers.tpl
Comment (errordeveloper) [inline install/kubernetes/cilium/templates/_helpers.tpl]: I don't understand this comment, can you please take a look and see if you can rephrase it? Seems like the reference to FQDN could be misused?

**Review by errordeveloper — COMMENTED** [review]
I don't know this enough to review it, let's try to ask someone who authored this part of the Helm chart. I think @kaworu or @gandro should be able to help :)

**Comment by gandro** [comment]
> I don't know this enough to review it, let's try to ask someone who authored this part of the Helm chart. I think @kaworu or @gandro should be able to help :)

Sounds good. I will take a look. Unassigned @kaworu as he is still out.

File: install/kubernetes/cilium/templates/_helpers.tpl
Comment (gandro) [inline install/kubernetes/cilium/templates/_helpers.tpl:90]: Please retain this comment where ever the `$_ := set . "ca" $ca` pattern is used.

File: install/kubernetes/cilium/templates/clustermesh-apiserver/tls-cronjob/cronjob.yaml
Comment (gandro) [inline install/kubernetes/cilium/templates/clustermesh-apiserver/tls-cronjob/cronjob.yaml]: This must not be removed. This is needed for `helm template -n NAMESPACE` to work correctly. See https://github.com/helm/helm/issues/3553#issuecomment-521424132

I locally tested this branch with helm 3.6.3 running `helm template --namespace kube-system` and the generated YAML was indeed missing the namespace in the cronjob.

**Note**: This is not the only place where you removed the namespace. To avoid unnecessary noise, I'll just point it out here and in the new files, but it applies to any resource touched by this PR.

File: install/kubernetes/cilium/templates/clustermesh-apiserver/tls-cronjob/job.yaml
Comment (gandro) [inline install/kubernetes/cilium/templates/clustermesh-apiserver/tls-cronjob/job.yaml:15]: This (and all other resources moved to a separate file) needs an explicit `namespace: {{ .Release.Namespace }}` as per the comment above.

Comment (gandro) [reply install/kubernetes/cilium/templates/clustermesh-apiserver/tls-cronjob/job.yaml:15]: But thanks for adding the checksum to this name here, btw!

File: install/kubernetes/cilium/templates/clustermesh-apiserver/tls-cronjob/_job-spec.tpl
Comment (gandro) [inline install/kubernetes/cilium/templates/clustermesh-apiserver/tls-cronjob/_job-spec.tpl:31]: I did a quick diff, and this flag is absent in the default output in the new version. I'm not familiar enough with external workloads, so it is possible that the previous version was broken, but I wonder the change is intended or not. I can take a closer look once the other issues are addressed.

File: install/kubernetes/cilium/templates/hubble/tls-helm/ca-secret.yaml
Comment (gandro) [inline install/kubernetes/cilium/templates/hubble/tls-helm/ca-secret.yaml]: This needs a check for `and (.Values.agent (not .Values.preflight.enabled))`. Otherwise, these secrets are created in the preflight check, which is wrong (as all other values are true here by default).

https://docs.cilium.io/en/v1.10/operations/upgrade/#running-pre-flight-check-required

You removed the `and (.Values.agent (not .Values.preflight.enabled))` check everywhere. I'm a bit nervous about that, we had issues in the past where the preflight check created artifacts that later broke the installation.

File: install/kubernetes/cilium/templates/hubble/tls-helm/relay-server-secret.yaml
Comment (gandro) [inline install/kubernetes/cilium/templates/hubble/tls-helm/relay-server-secret.yaml:1]: Same for this, this is created by default in the `preflight` check. We need to check for `and (.Values.agent (not .Values.preflight.enabled))` here.

**Review by gandro — CHANGES_REQUESTED** [review]
Thank you! I did a first pass based on what I spotted by running a few local diffs of `helm template` with different values.

File: install/kubernetes/cilium/templates/_helpers.tpl
Comment (dungdm93) [reply install/kubernetes/cilium/templates/_helpers.tpl:90]: This function is move to `templates/hubble/tls-helm/_helpers.tpl` because it's helm certgen specific.
I also keep the comment back, but change a little bit to match with new context. Could you please take a look @gandro

File: install/kubernetes/cilium/templates/clustermesh-apiserver/tls-cronjob/_job-spec.tpl
Comment (dungdm93) [reply install/kubernetes/cilium/templates/clustermesh-apiserver/tls-cronjob/_job-spec.tpl:31]: AFAIU, `client-cert` is used by externalWorkload only. 
So I make this change to align this with tls-provided and tls-helm certgen methods:
https://github.com/cilium/cilium/blob/901afa5131f27406c43a508356469141b6ed1ce0/install/kubernetes/cilium/templates/clustermesh-apiserver/client-secret.yaml#L1

File: install/kubernetes/cilium/templates/clustermesh-apiserver/tls-cronjob/cronjob.yaml
Comment (dungdm93) [reply install/kubernetes/cilium/templates/clustermesh-apiserver/tls-cronjob/cronjob.yaml]: `namespace` is back. Could you please take a look

File: install/kubernetes/cilium/templates/clustermesh-apiserver/tls-cronjob/job.yaml
Comment (dungdm93) [reply install/kubernetes/cilium/templates/clustermesh-apiserver/tls-cronjob/job.yaml:15]: `namespace` is back

File: install/kubernetes/cilium/templates/hubble/tls-helm/ca-secret.yaml
Comment (dungdm93) [reply install/kubernetes/cilium/templates/hubble/tls-helm/ca-secret.yaml]: Currently `and (.Values.agent (not .Values.preflight.enabled))` check is only apply to `tls-provided` and `tls-helm`.  
`tls-cronjob` is not using this check:
https://github.com/cilium/cilium/blob/901afa5131f27406c43a508356469141b6ed1ce0/install/kubernetes/cilium/templates/hubble/generate-certs-cronjob.yaml#L1
And deployment is not using as well:
https://github.com/cilium/cilium/blob/901afa5131f27406c43a508356469141b6ed1ce0/install/kubernetes/cilium/templates/hubble-relay/deployment.yaml#L1
Never the less, not all certs of `tls-provided` and `tls-helm` are using this check, named [`hubble-relay-client-certs`](https://github.com/cilium/cilium/blob/901afa5131f27406c43a508356469141b6ed1ce0/install/kubernetes/cilium/templates/hubble/relay-tls-client-secret.yaml#L1), [`hubble-relay-server-certs`](https://github.com/cilium/cilium/blob/901afa5131f27406c43a508356469141b6ed1ce0/install/kubernetes/cilium/templates/hubble/relay-tls-server-secret.yaml#L1) and [`hubble-ui-client-certs`](https://github.com/cilium/cilium/blob/901afa5131f27406c43a508356469141b6ed1ce0/install/kubernetes/cilium/templates/hubble/ui-tls-client-secret.yaml#L1)

So, my point is this check is not universal, my opinion this remove this and let `hubble.enabled` take care the rest.
WDYT, @gandro?

Comment (gandro) [reply install/kubernetes/cilium/templates/hubble/tls-helm/ca-secret.yaml]: > So, my point is this check is not universal, my opinion this remove this and let hubble.enabled take care the rest.
WDYT, @gandro?

The problem is that with the values we have in the docs, https://docs.cilium.io/en/v1.10/operations/upgrade/#running-pre-flight-check-required, users will have `hubble.enabled=true`. So just checking for `hubble.enabled` will not work with the existing documentation.

If users follow the documentation as it is right now, to run the preflight check, they will have the following values:

```ini
preflight.enabled=true # explicitly set by the user
agent=false # explicitly set by the user
operator.enabled=false  # explicitly set by the user
hubble.enabled=true # default value
hubble.tls.auto.enabled=true # default value
hubble.tls.auto.method="helm" # default value
hubble.relay.enabled=false # default value
```

This is why the `tls-helm` needs to check for `preflight.enabled` and `agent`, because all other values are default to true, so it will be applied by default (even though it shouldn't).

**Basically the rule is: If a component is enabled by default, then it needs to check for `preflight.enabled` to disable itself in preflight mode.** This is why `hubble-relay` or `tls-cronjob` currently do not have the preflight check, because they are not enabled by default.

But I fully agree that the inconsistency currently is not great. So let's fix it properly and as simple as possible.
My proposal would be the following.

- `hubble-server-certs` should only be created if `agent` is true.
   - Because it's only used by cilium-agent.
  -  In both modes: `tls-helm` and `tls-provided`.
  - The full check would be: `and .Values.agent .Values.hubble.enabled .Values.hubble.tls.enabled`
- `hubble-ca-secret` should only be created if `.agent` or  `hubble.relay.enabled` are true
  -  Because it's used by Agent or Relay, or UI (but UI requires Relay, so checking only for Relay is fine)
  -  In all three modes:`tls-helm`, `tls-provided`, `tls-cronjob`
  - The full check would be: `and (or .Values.agent .Values.hubble.relay.enabled) .Values.hubble.enabled .Values.hubble.tls.enabled`

The rest (`hubble-relay-server-certs`, `hubble-relay-client-certs`, `hubble-ui-client-certs`) already have to correct checks, they check for `relay` (which is disabled by default).

File: install/kubernetes/cilium/templates/hubble/tls-helm/_helpers.tpl
Comment (gandro) [inline install/kubernetes/cilium/templates/hubble/tls-helm/_helpers.tpl]: Thanks for adding back the comment. It needs to be adapted for the new system though:

```suggestion
Note: Always use this template as follows:

    {{- $_ := include "hubble-generate-certs.helm.setup-ca" . -}}

The assignment to `$_` is required because we store the generated CI in a global `ca` variable.
Please, don't try to "simplify" this, as without this trick, every generated
certificate would be signed by a different CA.
```

File: install/kubernetes/cilium/templates/clustermesh-apiserver/tls-helm/_helpers.tpl
Comment (gandro) [inline install/kubernetes/cilium/templates/clustermesh-apiserver/tls-helm/_helpers.tpl]:
```suggestion
Note: Always use this template as follows:

    {{- $_ := include "clustermesh-apiserver-generate-certs.helm.setup-ca" . -}}

The assignment to `$_` is required because we store the generated CI in a global `cmca` variable.
Please, don't try to "simplify" this, as without this trick, every generated
certificate would be signed by a different CA.
```

**Review by gandro — COMMENTED** [review]
Thank you for addressing the feedback, starting to look good. We still need to resolve the secrets in the preflight check, other than that just a small adjustment regarding the comment that you added back.

File: install/kubernetes/cilium/templates/_helpers.tpl
Comment (gandro) [reply install/kubernetes/cilium/templates/_helpers.tpl:90]: Thank you! I proposed a bit a more in-depth adjustment below.

File: install/kubernetes/cilium/templates/hubble/tls-helm/ca-secret.yaml
Comment (dungdm93) [reply install/kubernetes/cilium/templates/hubble/tls-helm/ca-secret.yaml]: Thank for your suggestion, it's updated.
I also checked `clustermesh` tls to match the logic, but I think you need to double check.

File: install/kubernetes/cilium/templates/hubble/tls-cronjob/ca-secret.yaml
Comment (gandro) [inline install/kubernetes/cilium/templates/hubble/tls-cronjob/ca-secret.yaml]: This line needs to be removed, as this secret unfortunately does not conform to the `tls` type. I got the following error while testing some upgrade paths on this branch:

```
Error: UPGRADE FAILED: cannot patch "hubble-ca-secret" with kind Secret: Secret "hubble-ca-secret" is invalid: [type: Invalid value: "kubernetes.io/tls": field is immutable, data[tls.crt]: Required value, data[tls.key]: Required value]
```

**Review by gandro — CHANGES_REQUESTED** [review]
Caught one more problem, the `kubernetes.io/tls` type cannot be used for the Hubble CA secret unfortunately.

File: install/kubernetes/cilium/templates/hubble/tls-cronjob/ca-secret.yaml
Comment (dungdm93) [reply install/kubernetes/cilium/templates/hubble/tls-cronjob/ca-secret.yaml]: My bad, it's updated.

**Review by gandro — APPROVED** [review]
Looks good now! Thanks a lot for your work :rocket:

**Comment by gandro** [comment]
test-me-please

**Comment by gandro** [comment]
net-next failed with an unrelated failure - https://jenkins.cilium.io/job/Cilium-PR-K8s-1.16-net-next/1133/testReport/junit/Suite-k8s-1/16/K8sServicesTest_Checks_service_across_nodes_Tests_NodePort_BPF_Tests_L2_less_with_Wireguard_provisioned_via_kube_wireguarder_Tests_NodePort_BPF/

I'll create an issue for it (it might actually be a new flake), but it is unrelated, as the cert logic is not affecting the failed test. No need to block this PR on this.

**Comment by dungdm93** [comment]
test-me-please

**Comment by dungdm93** [comment]
@gandro In the last commit, I add a release note about remove deprecated `hubble-ca-cert` configmap (which is noted that will be remove in cilium v1.11)
I also add a few changes in tls-cronjob as well:
* In `hubble/tls-cronjob/_job-spec.tpl`, no need to specify `*-cert-secret-name` and `*-cert-generate=false`, which are default options.
* In `clustermesh-apiserver/tls-cronjob/_job-spec.tpl` add support `validity-duration` options

File: Documentation/operations/upgrade.rst
Comment (qmonnet) [inline Documentation/operations/upgrade.rst:331]: Please re-add a blank line after this item and before the title of the next subsection.

**Review by qmonnet — APPROVED** [review]
I'm not super familiar with the Helm chart. I had a look at the changes and nothing particularly stood out, so all good for me.
Just one nit on the Documentation part, see comment below.

**Review by qmonnet — CHANGES_REQUESTED** [review]
[Marking as “Request changes” to leave you the time to address the nit before this gets merged]

**Comment by gandro** [comment]
Now that #16896 was merged, this now needs a rebase.

File: Documentation/operations/upgrade.rst
Comment (dungdm93) [reply Documentation/operations/upgrade.rst:331]: Blank line is added

**Comment by gandro** [comment]
test-me-please

**Review by qmonnet — APPROVED** [review]
Thanks!

**Comment by qmonnet** [comment]
test-me-please

**Comment by gandro** [comment]
Many GitHub Actions seems to have failed due to an issue with unavailable runner:
`This request was automatically failed because there were no enabled runners online to process the request for more than 1 days.`

https://github.com/cilium/cilium/actions?query=branch%3Arefactor-cert-gen-logic

I will try restart the failed ones. 

k8s-1.21-kernel-4.9 failed with some Hubble errors that I will triage in a bit.

File: install/kubernetes/cilium/templates/hubble/tls-helm/server-secret.yaml
Comment (gandro) [inline install/kubernetes/cilium/templates/hubble/tls-helm/server-secret.yaml]: Missing SANs:
```suggestion
{{- $cert := genSignedCert $cn nil (list $cn) (.Values.hubble.tls.auto.certValidityDuration | int) .ca -}}
```

File: install/kubernetes/cilium/templates/hubble/tls-helm/relay-server-secret.yaml
Comment (gandro) [inline install/kubernetes/cilium/templates/hubble/tls-helm/relay-server-secret.yaml]: Missing SANs:
```suggestion
{{- $cert := genSignedCert $cn nil (list $cn) (.Values.hubble.tls.auto.certValidityDuration | int) .ca -}}
```

File: install/kubernetes/cilium/templates/hubble/tls-helm/relay-client-secret.yaml
Comment (gandro) [inline install/kubernetes/cilium/templates/hubble/tls-helm/relay-client-secret.yaml]: Missing SANs:
```suggestion
{{- $cert := genSignedCert $cn nil (list $cn) (.Values.hubble.tls.auto.certValidityDuration | int) .ca -}}
```

File: install/kubernetes/cilium/templates/hubble/tls-helm/ui-client-certs.yaml
Comment (gandro) [inline install/kubernetes/cilium/templates/hubble/tls-helm/ui-client-certs.yaml]: Missing SANs:
```suggestion
{{- $cert := genSignedCert $cn nil (list $cn) (.Values.hubble.tls.auto.certValidityDuration | int) .ca -}}
```

**Review by gandro — CHANGES_REQUESTED** [review]
Thanks for the follow-ups. However, it seems like the failure in the Hubble test suite actually is legit:

```
2021-07-29T13:03:06.177107712Z level=warning msg="Failed to create gRPC client" address="192.168.36.11:4244" error="context deadline exceeded: connection error: desc = \"transport: authentication handshake failed: x509: certificate relies on legacy Common Name field, use SANs or temporarily enable Common Name matching with GODEBUG=x509ignoreCN=0\"" hubble-tls=true next-try-in=10m40s peer=k8s1 subsys=hubble-relay
```

I was able to reproduce it locally, by installing the Helm chart with Relay enabled against a local kind cluster:

```
export CILIUM_CI_TAG=$(git rev-parse HEAD)
helm upgrade --install cilium ./cilium \
  --namespace kube-system \
  --set image.repository=quay.io/cilium/cilium-ci \
  --set image.tag=$CILIUM_CI_TAG \
  --set operator.image.repository=quay.io/cilium/operator \
  --set operator.image.suffix=-ci \
  --set operator.image.tag=$CILIUM_CI_TAG \
  --set clustermesh.apiserver.image.repository=quay.io/cilium/clustermesh-apiserver-ci \
  --set clustermesh.apiserver.image.tag=$CILIUM_CI_TAG \
  --set hubble.relay.image.repository=quay.io/cilium/hubble-relay-ci \
  --set hubble.relay.image.tag=$CILIUM_CI_TAG \
  --set hubble.relay.enabled=true
```

I re-reviewed the cert sections and pointed out the missing SANs.

File: install/kubernetes/cilium/templates/hubble/tls-helm/relay-client-secret.yaml
Comment (dungdm93) [reply install/kubernetes/cilium/templates/hubble/tls-helm/relay-client-secret.yaml]: Actually client certificate is used to authenticate to server. It's no need to specify SANs.

**Comment by dungdm93** [comment]
@gandro Sorry for let you waiting, I quite busy last week.
I just change SANs as you suggested. But client certificate is used to authenticate to server, and its SANs is optional. So, do you think we need to specify SANs in `hubble-relay-client-certs` and `hubble-ui-client-certs`?

**Review by gandro — APPROVED** [review]

**Comment by gandro** [comment]
> @gandro Sorry for let you waiting, I quite busy last week.
> I just change SANs as you suggested. But client certificate is used to authenticate to server, and its SANs is optional. So, do you think we need to specify SANs in `hubble-relay-client-certs` and `hubble-ui-client-certs`?

No worries! That's a good question, my understanding is the same as you, i.e. that grpc/go only requires a SAN for server certs (https://pkg.go.dev/crypto/x509#Certificate.VerifyHostname). 

But since we are setting the SAN to CNs by default in `certgen` as well, it probably doesn't hurt to be consistent:
https://github.com/cilium/certgen/blob/7e5d372a03955949e544ce24ecd9ac16a9f00be7/internal/generate/generate.go#L70

**Comment by gandro** [comment]
test-me-please

**Comment by dungdm93** [comment]
@gandro `ConformanceEKS` test fails seem not related to this PR, could you rerun it.

**Comment by gandro** [comment]
> @gandro `ConformanceEKS` test fails seem not related to this PR, could you rerun it.

We can ignore the failure, rerunning it won't fix it unfortunately, as it is currently known to be broken: https://github.com/cilium/cilium/issues/17139

**Comment by aanm** [comment]
@gandro I assume we can merge this?

**Comment by gandro** [comment]
> @gandro I assume we can merge this?

@aanm There has be no further feedback by the `helm` subteam (which has yet to approve this PR), but I think we can merge it.

**Pull request merged**
